File: EZIN/Indicators/MovingAverages.swift
import Foundation

/// Moving averages — faithful port of utils/MovingAverages.jsx
/// (SMA, EMA, RMA, WMA, DEMA, TEMA, HMA, VWMA, KAMA).
enum MA {

    static func sma(_ src: [Double], _ len: Int) -> [Double] {
        src.indices.map { i in
            guard i >= len - 1 else { return 0 }
            var s = 0.0
            for j in (i - len + 1)...i { s += src[j] }
            return s / Double(len)
        }
    }

    static func ema(_ src: [Double], _ len: Int) -> [Double] {
        let k = 2.0 / Double(len + 1)
        var out = [Double]()
        for (i, v) in src.enumerated() {
            out.append(i == 0 ? v : v * k + out[i - 1] * (1 - k))
        }
        return out
    }

    /// Wilder's smoothing (RMA) — used by RSI / ATR.
    static func rma(_ src: [Double], _ len: Int) -> [Double] {
        let alpha = 1.0 / Double(len)
        var out = [Double]()
        for (i, v) in src.enumerated() {
            out.append(i == 0 ? v : alpha * v + (1 - alpha) * out[i - 1])
        }
        return out
    }

    static func wma(_ src: [Double], _ len: Int) -> [Double] {
        src.indices.map { i in
            guard i >= len - 1 else { return 0 }
            var num = 0.0, den = 0.0
            for j in 0..<len { num += src[i - j] * Double(len - j); den += Double(len - j) }
            return num / den
        }
    }

    static func dema(_ src: [Double], _ len: Int) -> [Double] {
        let e1 = ema(src, len), e2 = ema(e1, len)
        return src.indices.map { 2 * e1[$0] - e2[$0] }
    }

    static func tema(_ src: [Double], _ len: Int) -> [Double] {
        let e1 = ema(src, len), e2 = ema(e1, len), e3 = ema(e2, len)
        return src.indices.map { 3 * e1[$0] - 3 * e2[$0] + e3[$0] }
    }

    static func hma(_ src: [Double], _ len: Int) -> [Double] {
        let half = len / 2
        let sqrtLen = Int(Double(len).squareRoot())
        let w1 = wma(src, max(half, 1)), w2 = wma(src, len)
        let raw = src.indices.map { 2 * w1[$0] - w2[$0] }
        return wma(raw, max(sqrtLen, 1))
    }

    static func vwma(_ src: [Double], _ vol: [Double], _ len: Int) -> [Double] {
        src.indices.map { i in
            guard i >= len - 1 else { return 0 }
            var num = 0.0, den = 0.0
            for j in (i - len + 1)...i { num += src[j] * vol[j]; den += vol[j] }
            return den != 0 ? num / den : 0
        }
    }

    static func kama(_ src: [Double], _ len: Int = 10, _ fast: Int = 2, _ slow: Int = 30) -> [Double] {
        let fastSC = 2.0 / Double(fast + 1), slowSC = 2.0 / Double(slow + 1)
        var out = [Double]()
        for (i, v) in src.enumerated() {
            if i < len { out.append(v); continue }
            let change = abs(v - src[i - len])
            var vol = 0.0
            for j in (i - len + 1)...i { vol += abs(src[j] - src[j - 1]) }
            let er = vol != 0 ? change / vol : 0
            let sc = pow(er * (fastSC - slowSC) + slowSC, 2)
            out.append(out[i - 1] + sc * (v - out[i - 1]))
        }
        return out
    }
}

File: src/indicators/movingAverages.ts
// Moving averages — ported from EZIN/Indicators/MovingAverages.swift
// (SMA, EMA, RMA, WMA, DEMA, TEMA, HMA, VWMA, KAMA).

export function sma(src: number[], len: number): number[] {
  return src.map((_, i) => {
    if (i < len - 1) return 0
    let s = 0
    for (let j = i - len + 1; j <= i; j++) s += src[j]
    return s / len
  })
}

export function ema(src: number[], len: number): number[] {
  const k = 2 / (len + 1)
  const out: number[] = []
  for (let i = 0; i < src.length; i++) {
    out.push(i === 0 ? src[i] : src[i] * k + out[i - 1] * (1 - k))
  }
  return out
}

/** Wilder's smoothing (RMA) — used by RSI / ATR. */
export function rma(src: number[], len: number): number[] {
  const alpha = 1 / len
  const out: number[] = []
  for (let i = 0; i < src.length; i++) {
    out.push(i === 0 ? src[i] : alpha * src[i] + (1 - alpha) * out[i - 1])
  }
  return out
}

export function wma(src: number[], len: number): number[] {
  return src.map((_, i) => {
    if (i < len - 1) return 0
    let num = 0
    let den = 0
    for (let j = 0; j < len; j++) {
      num += src[i - j] * (len - j)
      den += len - j
    }
    return num / den
  })
}

export function dema(src: number[], len: number): number[] {
  const e1 = ema(src, len)
  const e2 = ema(e1, len)
  return src.map((_, i) => 2 * e1[i] - e2[i])
}

export function tema(src: number[], len: number): number[] {
  const e1 = ema(src, len)
  const e2 = ema(e1, len)
  const e3 = ema(e2, len)
  return src.map((_, i) => 3 * e1[i] - 3 * e2[i] + e3[i])
}

export function hma(src: number[], len: number): number[] {
  const half = Math.floor(len / 2)
  const sqrtLen = Math.floor(Math.sqrt(len))
  const w1 = wma(src, Math.max(half, 1))
  const w2 = wma(src, len)
  const raw = src.map((_, i) => 2 * w1[i] - w2[i])
  return wma(raw, Math.max(sqrtLen, 1))
}

export function vwma(src: number[], vol: number[], len: number): number[] {
  return src.map((_, i) => {
    if (i < len - 1) return 0
    let num = 0
    let den = 0
    for (let j = i - len + 1; j <= i; j++) {
      num += src[j] * vol[j]
      den += vol[j]
    }
    return den !== 0 ? num / den : 0
  })
}

export function kama(src: number[], len = 10, fast = 2, slow = 30): number[] {
  const fastSC = 2 / (fast + 1)
  const slowSC = 2 / (slow + 1)
  const out: number[] = []
  for (let i = 0; i < src.length; i++) {
    const v = src[i]
    if (i < len) {
      out.push(v)
      continue
    }
    const change = Math.abs(v - src[i - len])
    let vol = 0
    for (let j = i - len + 1; j <= i; j++) vol += Math.abs(src[j] - src[j - 1])
    const er = vol !== 0 ? change / vol : 0
    const sc = Math.pow(er * (fastSC - slowSC) + slowSC, 2)
    out.push(out[i - 1] + sc * (v - out[i - 1]))
  }
  return out
}
